### PR TITLE
[FLINK-16029][table-planner-blink] Remove register source and sink in test cases of blink planner

### PR DESCRIPTION
**Pull request by docete**
## What is the purpose of the change
Many test cases of planner use TableEnvironement.registerTableSource() and registerTableSink() which should be avoid. We want to refactor these cases via TableEnvironment.connect(). This PR 
remove most of these calls except in two situations

- table sources implements DefinedRowtimeAttributes or DefinedProctimeAttribute, which will be handled by https://issues.apache.org/jira/browse/FLINK-16160. These sources are： TestTableSourceWithTime/TestPreserveWMTableSource/TestProjectableTableSource/TestNestedProjectableTableSource

-  TableTestBase#addTableSource which will be handled by https://issues.apache.org/jira/browse/FLINK-16117

## Brief change log

- b53f6b1 Port CustomConnectorDescriptor to flink-table-common module
- f38976 Replace TableEnvironment.registerTableSource/Sink() by TableEnvironment.connect()

## Verifying this change

This change is already covered by existing tests.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (yes / **no**)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (yes / **no**)
  - The serializers: (yes / **no** / don't know)
  - The runtime per-record code paths (performance sensitive): (yes / **no** / don't know)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: (yes / **no** / don't know)
  - The S3 file system connector: (yes / **no** / don't know)

## Documentation

  - Does this pull request introduce a new feature? (**yes** / no)
  - If yes, how is the feature documented? (not applicable / docs / **JavaDocs** / not documented)

### CODE REVIEW

File: flink-table/flink-table-planner-blink/src/test/scala/org/apache/flink/table/planner/runtime/stream/sql/AsyncLookupJoinITCase.scala [hunk 1]
```diff
@@ -99,7 +93,6 @@ class AsyncLookupJoinITCase(backend: StateBackendMode)
       "1,12,Julian",
       "3,15,Fabian")
     assertEquals(expected.sorted, sink.getAppendResults.sorted)
-    assertEquals(0, userTableSourceWith2Keys.getResourceCounter)
```
Comment (godfreyhe) [inline flink-table/flink-table-planner-blink/src/test/scala/org/apache/flink/table/planner/runtime/stream/sql/AsyncLookupJoinITCase.scala]: can't get `userTableSourceWith2Keys.getResourceCounter` using `tEnv.connect` ?

Comment (docete) [reply flink-table/flink-table-planner-blink/src/test/scala/org/apache/flink/table/planner/runtime/stream/sql/AsyncLookupJoinITCase.scala]: We can't fetch the actual table source `userTableSourceWith2Keys` if we use tEnv.connect or DDL. Thus we can't verify the counter in that table source instance.

**Comment by docete** [comment]
wrapped 'tableEnv.connect()' in testing utilities of user-defined table sources.

File: flink-table/flink-table-planner-blink/src/test/scala/org/apache/flink/table/planner/utils/testTableSources.scala [hunk 2]
```diff
@@ -705,6 +724,63 @@ class TestDataTypeTableSource(
...
+
+  override def supportedProperties(): JList[String] = {
+    val supported = new util.ArrayList[String]()
+    supported.add("*")
+    supported
+  }
+}
+
+object TestDataTypeTableSource {
+  var data: Seq[Row] = null
```
Comment (godfreyhe) [inline flink-table/flink-table-planner-blink/src/test/scala/org/apache/flink/table/planner/utils/testTableSources.scala:753]: avoid `static variable`

File: flink-table/flink-table-planner-blink/src/test/scala/org/apache/flink/table/planner/runtime/stream/table/TableSinkITCase.scala [hunk 3]
```diff
@@ -26,54 +26,19 @@ import org.apache.flink.table.api.scala._
...
-
 import java.io.File
 import java.util.TimeZone
 
 import scala.collection.JavaConverters._
 
 class TableSinkITCase extends AbstractTestBase {
 
-  @Test
-  def testInsertIntoRegisteredTableSink(): Unit = {
```
Comment (godfreyhe) [inline flink-table/flink-table-planner-blink/src/test/scala/org/apache/flink/table/planner/runtime/stream/table/TableSinkITCase.scala:46]: is this test unnecessary ?

File: flink-table/flink-table-planner-blink/src/test/scala/org/apache/flink/table/api/TableEnvironmentITCase.scala [hunk 4]
```diff
@@ -141,7 +154,6 @@ class TableEnvironmentITCase(tableEnvName: String, isStreaming: Boolean) {
 
     val result1 = tEnv.explain(false)
     val result2 = tEnv.explain(false)
-    assertEquals(replaceStageId(result1), replaceStageId(result2))
```
Comment (godfreyhe) [inline flink-table/flink-table-planner-blink/src/test/scala/org/apache/flink/table/api/TableEnvironmentITCase.scala:144]: revert this?

File: flink-table/flink-table-api-java-bridge/src/main/java/org/apache/flink/table/descriptors/OldCsv.java [hunk 5]
```diff
@@ -228,6 +232,25 @@ public OldCsv ignoreParseErrors() {
 		return this;
 	}
 
+	/**
```
Comment (godfreyhe) [inline flink-table/flink-table-api-java-bridge/src/main/java/org/apache/flink/table/descriptors/OldCsv.java:235]: move these changes to a separate commit

File: flink-table/flink-table-planner-blink/src/test/scala/org/apache/flink/table/planner/runtime/utils/InMemoryLookupableTableSource.scala [hunk 6]
```diff
@@ -73,121 +79,72 @@ class InMemoryLookupableTableSource(
...
+  override def supportedProperties(): util.List[String] = {
+    val supported = new util.ArrayList[String]()
+    supported.add("*")
+    supported
+  }
+}
+
+object InMemoryLookupableTableSource {
+
+  def createTemporaryTable(
```
[the remark this answers is not in the document]
Comment (docete) [reply flink-table/flink-table-planner-blink/src/test/scala/org/apache/flink/table/planner/runtime/utils/InMemoryLookupableTableSource.scala:134]: Using `createTemporaryTable` doesn't mean that it will return something. See `TableEnvironment.createTemporaryView` or `ConnectorTableDescriptor.createTemporaryTable`. I think we should use an unify naming.

File: flink-table/flink-table-planner-blink/src/test/scala/org/apache/flink/table/planner/utils/testTableSources.scala [hunk 7]
```diff
@@ -54,8 +55,67 @@ import scala.collection.JavaConverters._
 import scala.collection.mutable
 
 object TestTableSources {
+  def createPersonCsvTemporaryTable(tEnv: TableEnvironment, tableName: String): Unit = {
```
[the remark this answers is not in the document]
Comment (docete) [reply flink-table/flink-table-planner-blink/src/test/scala/org/apache/flink/table/planner/utils/testTableSources.scala:58]: ditto

File: flink-table/flink-table-planner-blink/src/test/scala/org/apache/flink/table/planner/runtime/stream/table/TableSinkITCase.scala (hunk 3 above)
Comment (docete) [reply flink-table/flink-table-planner-blink/src/test/scala/org/apache/flink/table/planner/runtime/stream/table/TableSinkITCase.scala:46]: There is no registered table sink any more.

File: flink-table/flink-table-planner-blink/src/test/scala/org/apache/flink/table/planner/utils/MemoryTableSourceSinkUtil.scala [hunk 8]
```diff
@@ -112,29 +78,28 @@ object MemoryTableSourceSinkUtil {
...
-    override def configure(
-        fieldNames: Array[String],
-        fieldTypes: Array[TypeInformation[_]]): TableSink[Row] = {
-      val newSink = new UnsafeMemoryOutputFormatTableSink
-      newSink.fieldNames = fieldNames
-      newSink.fieldTypes = fieldTypes
-      newSink
+    override def requiredContext(): util.Map[String, String] = {
+      val context = new util.HashMap[String, String]()
+      context.put(CONNECTOR_TYPE, "LegacyUnsafeMemoryAppendTable")
```
[the remark this answers is not in the document]
Comment (docete) [reply flink-table/flink-table-planner-blink/src/test/scala/org/apache/flink/table/planner/utils/MemoryTableSourceSinkUtil.scala:94]: It's not necessary to be a constant field, it's only used in this test case.

File: flink-table/flink-table-planner-blink/src/test/scala/org/apache/flink/table/planner/plan/stream/sql/TableSourceTest.scala [hunk 9]
```diff
@@ -39,34 +36,23 @@ class TableSourceTest extends TableTestBase {
...
-    val tableSource = new TableSource[Row]() {
-
-      override def getProducedDataType: DataType = tableSchema.toRowDataType
-
-      override def getTableSchema: TableSchema = tableSchema
-    }
-    util.tableEnv.registerTableSource("MyTable", tableSource)
-    thrown.expect(classOf[ValidationException])
-    thrown.expectMessage(
-      "Only StreamTableSource and LookupableTableSource can be used in Blink planner.")
```
[the remark this answers is not in the document]
Comment (docete) [reply flink-table/flink-table-planner-blink/src/test/scala/org/apache/flink/table/planner/plan/stream/sql/TableSourceTest.scala:69]: It's only used in `convertSourceTable` which convert `ConnectorCatalogTable` to `TableSourceTable`. When there is no registerTableSource/Sink, there is no `ConnectorCatalogTable` either.

**Comment by godfreyhe** [comment]
LGTM

**Comment by docete** [comment]
rebase. 
@wuchong Could you take a look and merge this?

**Comment by docete** [comment]
@flinkbot run travis

**Comment by KurtYoung** [comment]
@docete thanks for working on this. Please resolve the conflicts and I will give a final pass and then merge it

**Comment by docete** [comment]
rebased to resolve conflicts.
@KurtYoung please have a look.

File: flink-table/flink-table-api-java-bridge/src/main/java/org/apache/flink/table/sinks/CsvTableSinkFactoryBase.java [hunk 10]
```diff
@@ -77,7 +81,12 @@
 		properties.add(FORMAT_FIELDS + ".#." + DescriptorProperties.TABLE_SCHEMA_NAME);
 		properties.add(FormatDescriptorValidator.FORMAT_DERIVE_SCHEMA);
 		properties.add(FORMAT_FIELD_DELIMITER);
+		properties.add(FORMAT_WRITE_MODE);
+		properties.add(FORMAT_NUM_FILES);
```
Comment (KurtYoung) [inline flink-table/flink-table-api-java-bridge/src/main/java/org/apache/flink/table/sinks/CsvTableSinkFactoryBase.java:85]: duplicated

File: flink-table/flink-table-planner-blink/src/test/scala/org/apache/flink/table/api/TableEnvironmentITCase.scala [hunk 11]
```diff
@@ -72,13 +69,20 @@ class TableEnvironmentITCase(tableEnvName: String, isStreaming: Boolean) extends
...
-    tEnv.registerTableSource("MyTable", getPersonCsvTableSource)
+    TestTableSourceSinks.createPersonCsvTemporaryTable(tEnv, "MyTable")
   }
 
   @Test
   def testExecuteTwiceUsingSameTableEnv(): Unit = {
-    val sink1Path = registerCsvTableSink(tEnv, Array("first"), Array(STRING), "MySink1")
-    val sink2Path = registerCsvTableSink(tEnv, Array("last"), Array(STRING), "MySink2")
+    val sink1Path = _tempFolder.newFile().getAbsolutePath
+    TestTableSourceSinks.createCsvTemporarySinkTable(
```
Comment (KurtYoung) [inline flink-table/flink-table-planner-blink/src/test/scala/org/apache/flink/table/api/TableEnvironmentITCase.scala:78]: We could let `createCsvTemporarySinkTable` method create the temporary path and then return it

File: flink-table/flink-table-planner-blink/src/test/scala/org/apache/flink/table/planner/runtime/batch/sql/TimestampITCase.scala [hunk 12]
```diff
@@ -41,12 +40,12 @@ class TimestampITCase extends BatchTestBase {
...
-        DataTypes.TIMESTAMP(3).bridgedTo(classOf[LocalDateTime]),
-        DataTypes.TIMESTAMP(3).bridgedTo(classOf[Timestamp]),
-        DataTypes.TIMESTAMP_WITH_LOCAL_TIME_ZONE(9).bridgedTo(classOf[Instant]),
-        DataTypes.TIMESTAMP_WITH_LOCAL_TIME_ZONE(9).bridgedTo(classOf[Instant]))
+        DataTypes.TIMESTAMP(9),
+        DataTypes.TIMESTAMP(9),
+        DataTypes.TIMESTAMP(3),
+        DataTypes.TIMESTAMP(3),
+        DataTypes.TIMESTAMP_WITH_LOCAL_TIME_ZONE(9),
+        DataTypes.TIMESTAMP_WITH_LOCAL_TIME_ZONE(9))
```
Comment (KurtYoung) [inline flink-table/flink-table-planner-blink/src/test/scala/org/apache/flink/table/planner/runtime/batch/sql/TimestampITCase.scala:48]: Why change these?

Comment (docete) [reply flink-table/flink-table-planner-blink/src/test/scala/org/apache/flink/table/planner/runtime/batch/sql/TimestampITCase.scala:48]: specifying conversion class of the types is not allowed in DDL and Descriptor API.
